opamp-client/src/main/java/io/opentelemetry/opamp/client/internal/request/service/HttpRequestService.java

```diff
@@ -0,0 +1,225 @@
+    try {
+      AgentToServer agentToServer =
+          Objects.requireNonNull(requestSupplier).get().getAgentToServer();
+
+      byte[] data = agentToServer.encodeByteString().toByteArray();
+      try (HttpSender.Response response =
+          requestSender.send(new ByteArrayWriter(data), data.length).get()) {
+        if (isSuccessful(response)) {
+          handleSuccessResponse(
+              Response.create(ServerToAgent.ADAPTER.decode(response.bodyInputStream())));
+        } else {
+          handleHttpError(response);
+        }
+      } catch (IOException e) {
+        getCallback().onRequestFailed(e);
+      }
+
+    } catch (InterruptedException e) {
+      getCallback().onRequestFailed(e);
+    } catch (ExecutionException e) {
+      if (e.getCause() != null) {
+        getCallback().onRequestFailed(e.getCause());
+      } else {
+        getCallback().onRequestFailed(e);
+      }
+    }
+  }
+
+  private void handleHttpError(HttpSender.Response response) {
+    int errorCode = response.statusCode();
+    getCallback().onRequestFailed(new HttpErrorException(errorCode, response.statusMessage()));
+
+    if (errorCode == 503 || errorCode == 429) {
+      String retryAfterHeader = response.getHeader("Retry-After");
+      Duration retryAfter = null;
+      if (retryAfterHeader != null) {
+        Optional<Duration> duration = retryAfterParser.tryParse(retryAfterHeader);
+        if (duration.isPresent()) {
+          retryAfter = duration.get();
+        }
+      }
+      enableRetryMode(retryAfter);
+    }
+  }
+
+  private static boolean isSuccessful(HttpSender.Response response) {
+    return response.statusCode() >= 200 && response.statusCode() < 300;
+  }
+
+  private void handleSuccessResponse(Response response) {
+    if (retryModeEnabled.get()) {
+      disableRetryMode();
+    }
+    ServerToAgent serverToAgent = response.getServerToAgent();
+
+    if (serverToAgent.error_response != null) {
+      handleErrorResponse(serverToAgent.error_response);
+    }
+
+    getCallback().onRequestSuccess(response);
+  }
+
+  private void handleErrorResponse(ServerErrorResponse errorResponse) {
+    if (errorResponse.type.equals(ServerErrorResponseType.ServerErrorResponseType_Unavailable)) {
+      Duration retryAfter = null;
+      if (errorResponse.retry_info != null) {
+        retryAfter = Duration.ofNanos(errorResponse.retry_info.retry_after_nanoseconds);
+      }
+      enableRetryMode(retryAfter);
+    }
+  }
+
+  private Callback getCallback() {
+    return Objects.requireNonNull(callback);
+  }
+
+  private static class ByteArrayWriter implements Consumer<OutputStream> {
+    private final byte[] data;
+
+    private ByteArrayWriter(byte[] data) {
+      this.data = data;
+    }
+
+    @SuppressWarnings("ThrowSpecificExceptions")
+    @Override
+    public void accept(OutputStream outputStream) {
+      try {
+        outputStream.write(data);
+      } catch (IOException e) {
+        throw new RuntimeException(e);
```

Comment:
You could use another exception here to avoid the need to suppress the warning. Though you could also consider whether `Consumer<OutputStream>` is what you want to use. Since `HttpSender` is an internal api and the only user of it provides `byte[]` then perhaps that is what would be the easiest, or you could create a custom `ThrowingConsumer` class that would let you avoid handling the exception here.

Reply:
I'm thinking about making HttpSender public for whenever this lib becomes stable, because I'm guessing some people would want to create their own implementations that won't depend on okhttp (as I'm aware of some issues we've had in core with people not linking the kotlin transitive dependency that it brings), which is why I'd like to keep it as generic as possible. So I've created an interface named `BodyWriter` that is more explicit about what it does and also throws an exception so that it's handled on the caller side.